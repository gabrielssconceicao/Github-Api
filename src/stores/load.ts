import { ref } from 'vue';
import { defineStore } from 'pinia';

export const useLoadStore = defineStore('load', () => {
  const loading = ref(false);
  function setLoading() {
    loading.value = true;
  }
  function setNotLoading() {
    loading.value = false;
  }
  return { loading, setLoading, setNotLoading };
});
