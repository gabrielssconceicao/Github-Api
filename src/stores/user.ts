import { ref } from 'vue';
import { defineStore } from 'pinia';
import type { IUserResponse } from '@/types/user';
import { fetchUser } from '@/hooks/fetchUser';
export const useUserStore = defineStore('counter', () => {
  const githubUser = ref<IUserResponse>({} as IUserResponse);

  async function searchUser(userLogin: string) {
    const data = await fetchUser(userLogin);
    githubUser.value = data;
  }

  return { githubUser, searchUser };
});
