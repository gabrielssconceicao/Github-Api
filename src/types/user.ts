export interface IUser {
  name: null | string;
  login: string;
  avatar_url: string;
  followers: number;
  following: number;
  location: null | string;
  html_url: string;
  bio: null | string;
  public_repos: number;
  repos_url: string;
}

export interface IUserResponse {
  user: IUser;
  status: number;
}
