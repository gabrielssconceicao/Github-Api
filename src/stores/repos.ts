import { defineStore } from 'pinia';
import { ref, computed } from 'vue';

import { type IRepos } from '@/types/repos';

import { langSeparator } from '@/helpers/langSeparator';
import { useLoadStore } from './load';
export const useReposStore = defineStore('repos', () => {
  const repos = ref<IRepos[]>([]);
  const reposUrl = ref('');
  const filterValue = ref('');

  function setRepos(reposValue: IRepos[]) {
    repos.value = reposValue;
  }

  const setFilter = (value: string) => {
    filterValue.value = value;
  };

  function setReposUrl(url: string) {
    reposUrl.value = url;
  }
  async function getRepos() {
    const useLoad = useLoadStore();
    useLoad.setLoading();
    try {
      const response = await fetch(reposUrl.value);
      const data: IRepos[] = await response.json();
      const dataWithoutNull = data.map((e) => {
        if (e.language) return e;
        return { ...e, language: 'No Language' };
      });
      setRepos(dataWithoutNull);
    } catch (error) {
      setRepos([]);
    } finally {
      useLoad.setNotLoading();
    }
  }

  const filteredRepos = computed(() => {
    return filterValue.value !== ''
      ? repos.value.filter((repo) => repo.language === filterValue.value)
      : repos.value;
  });

  const filteredReposCount = computed(
    () => repos.value && langSeparator(repos.value),
  );

  return {
    filteredRepos,
    getRepos,
    filteredReposCount,
    setFilter,
    setReposUrl,
    setRepos,
    repos,
    reposUrl,
    filterValue,
  };
});
