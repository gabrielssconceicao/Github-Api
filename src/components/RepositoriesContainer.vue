<script setup lang="ts">

import ReposHeader from './ReposHeader.vue';
import ReposFilter from './ReposFilterComponent.vue';
import ReposComponent from './ReposComponent.vue';

import { useUserStore } from '@/stores/user'
import { useReposStore } from '@/stores/repos'
const useUser = useUserStore()
const useRepos = useReposStore()


</script>

<template>
  <ReposHeader :name="useUser.userName" />

  <section class="h-[calc(100dvh-51.2px)] overflow-y-auto  relative" id="repos">

    <ReposFilter :filters="useRepos.filteredReposCount" @filter="useRepos.setFilter" class="sticky top-0 z-10" />
    <div class="max-w-7xl mx-auto overflow-y-auto py-8 ">
      <TransitionGroup name="fade" tag="div" class="grid gap-3 px-2  grid-cols-1 justify-items-stretch"
        id="repos-component">
        <ReposComponent v-for="repo in useRepos.filteredRepos" :key="repo.id" :repos="repo" />
      </TransitionGroup>
    </div>
  </section>
</template>
<style scoped>
#repos-component {
  grid-template-columns: repeat(auto-fill, minmax(275px, 1fr));
  grid-template-rows: repeat(auto-fill, minmax(150px, 1fr));
  align-items: stretch;

}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s ease-in-out;
}

.fade-enter-from,
.fade-leave-to {
  opacity: 0;
}
</style>
