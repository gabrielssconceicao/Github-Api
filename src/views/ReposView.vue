<script setup lang="ts">
import RepositoriesContainer from '@/components/RepositoriesContainer.vue';
import Loading from '@/components/LoadingComponent.vue'
import { useLoadStore } from '@/stores/load'
const useLoad = useLoadStore()
</script>
<template>
  <RepositoriesContainer v-if="!useLoad.loading" />
  <Loading v-if="useLoad.loading" />
</template>


