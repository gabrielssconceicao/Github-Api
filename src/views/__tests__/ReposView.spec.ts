import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import ReposView from '../ReposView.vue';
import Loading from '@/components/LoadingComponent.vue';
import RepositoriesContainer from '@/components/RepositoriesContainer.vue';
import { useUserStore } from '@/stores/user';
describe('<ReposView />', () => {
  it('should render correctly', async () => {
    const useUser = useUserStore();
    useUser.setInputValue('resolveTest');
    useUser.searchUser();
    const wrapper = mount(ReposView);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.findComponent(Loading).exists()).toBe(true);
    expect(wrapper.findComponent(RepositoriesContainer).exists()).toBe(false);
    await wrapper.vm.$nextTick();
    expect(wrapper.findComponent(Loading).exists()).toBe(true);
    expect(wrapper.findComponent(RepositoriesContainer).exists()).toBe(false);
  });

  it('should match snapshot', async () => {
    const useUser = useUserStore();
    useUser.setInputValue('resolveTest');
    await useUser.searchUser();
    const wrapper = mount(ReposView);

    expect(wrapper.element).toMatchSnapshot();
  });
});
