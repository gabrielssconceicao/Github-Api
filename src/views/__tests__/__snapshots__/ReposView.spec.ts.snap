// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ReposView /> > should match snapshot 1`] = `
<div
  data-v-app=""
>
  
  
  <header
    class="bg-header p-2 flex items-center text-white"
    data-v-f09d8ef3=""
  >
    <routerlink
      class="w-14 border-white border-2 p-1 flex justify-center items-center"
      to="/"
    >
      Back 
    </routerlink>
    <p
      class="flex-1 text-center"
    >
      <span
        class="font-bold font-mono text-xl text-primary"
      >
        Teste User
      </span>
       Repos 
    </p>
  </header>
  <section
    class="h-[calc(100dvh-51.2px)] overflow-y-auto relative"
    data-v-f09d8ef3=""
    id="repos"
  >
    <section
      class="flex flex-col items-center justify-center flex-1 gap-3 font-mono px-1 h-[inherit]"
      data-v-f09d8ef3=""
    >
      <h2
        class="text-xl md:text-3xl font-bold text-red-600 text-center"
      >
        This User has no Repositories 
      </h2>
    </section>
  </section>
  
  <!--v-if-->
  <!--v-if-->
  
</div>
`;
